Error in sql call from REST API

```
[2019-09-12T20:16:50.491] [ERROR] default - getOneTransaction error: {"stack":"error: bind message supplies 5 parameters, but prepared statement \"\" requires 4\n at Connection.parseE (/opt/restapi/rest-api/node_modules/pg/lib/connection.js:604:11)\n at Connection.parseMessage (/opt/restapi/rest-api/node_modules/pg/lib/connection.js:401:19)\n at Socket.<anonymous> (/opt/restapi/rest-api/node_modules/pg/lib/connection.js:121:22)\n at Socket.emit (events.js:209:13)\n at addChunk (_stream_readable.js:305:12)\n at readableAddChunk (_stream_readable.js:286:11)\n at Socket.Readable.push (_stream_readable.js:220:10)\n at TCP.onStreamRead (internal/stream_base_commons.js:182:23)","message":"bind message supplies 5 parameters, but prepared statement \"\" requires 4","name":"error","length":137,"severity":"ERROR","code":"08P01","file":"postgres.c","line":"1559","routine":"exec_bind_message"}
```

--- a/README.md
+++ b/README.md
@@ -250,7 +250,7 @@ Create a `.env` file as per below and run with `npm start`.
 
 You can also unittest using jest by using `npm test`.
 
-example `.env` file:
+example `.env` file (note: there is a env.sample file in this directory):
 
 ```TEXT
 DB_USER=mirror_api
@@ -260,6 +260,12 @@ DB_NAME=mirror_node
 PORT=5551
 # server hosting the database
 DB_HOST=localhost
+# Shard number
+SHARD_NUM=0
+# Whether to include the host:port in the next links returned from the rest api
+INCLUDE_PATH_IN_NEXT_LINKS=0
+
+
 ```
 
 `PORT` is the port number the REST API will listen onto.

--- a/rest-api/accounts.js
+++ b/rest-api/accounts.js
@@ -95,7 +95,7 @@ const getAccounts = function (req) {
         ")\n";
 
     const [balanceQuery, balanceParams] = utils.parseParams(req, 'account.balance',
-        ['ab.balance']);
+        ['ab.balance'], 'balance');
 
     let [pubKeyQuery, pubKeyParams] = utils.parseParams(req, 'account.publickey',
         ['e.ed25519_public_key_hex'], null, (s) => {return s.toLowerCase();});

--- a/rest-api/balances.js
+++ b/rest-api/balances.js
@@ -55,7 +55,7 @@ const getBalances = function (req) {
         'timestamp', ['ab.consensus_timestamp'], 'timestamp_ns');
 
     let [balanceQuery, balanceParams] = utils.parseParams(req, 'account.balance',
-        ['ab.balance']);
+        ['ab.balance'], 'balance');
 
     let [pubKeyQuery, pubKeyParams] = utils.parseParams(req, 'account.publickey',
         ['e.ed25519_public_key_hex'], null, (s) => {return s.toLowerCase();});

--- a/rest-api/env.sample
+++ b/rest-api/env.sample
@@ -4,3 +4,4 @@ DB_USER='mirror_api'
 DB_PASS='mirror_api_pass'
 DB_NAME='mirror_node'
 SHARD_NUM=0
+INCLUDE_PATH_IN_NEXT_LINKS=0

--- a/rest-api/transactions.js
+++ b/rest-api/transactions.js
@@ -249,7 +249,19 @@ const getOneTransaction = function (req, res) {
 
     // The transaction id is in the format of 'shard.realm.num-ssssssssss.nnnnnnnnn'
     // convert it in shard, realm, num and nanoseconds parameters
-    const sqlParams = req.params.id.replace(/(\d{10})-(\d{9})/, '$1$2').split(/[.-]/);
+    let txIdMatches = req.params.id.match(/(\d+)\.(\d+)\.(\d+)-(\d*)-(\d*)/);
+    if (txIdMatches === null || txIdMatches.length != 6) {
+        logger.info(`getOneTransaction: Invalid transaction id ${req.params.id}`);
+        res.status(404)
+            .send('Invalid Transaction id. Please use "shard.realm.num-ssssssssss.nnnnnnnnn" ' +
+                'format where ssss are 10 digits seconds and nnn are 9 digits nanoseconds');
+        return;
+    }
+    const sqlParams = [
+        txIdMatches[1], txIdMatches[2], txIdMatches[3], 
+        txIdMatches[4] + '' + txIdMatches[5]
+    ];
+
 
     let sqlQuery =
         "select etrans.entity_shard,  etrans.entity_realm, etrans.entity_num\n" +

--- a/rest-api/transactions.js
+++ b/rest-api/transactions.js
@@ -247,9 +247,9 @@ const getOneTransaction = function (req, res) {
     logger.debug("--------------------  getTransactions --------------------");
     logger.debug("Client: [" + req.ip + "] URL: " + req.originalUrl);
 
-    // The transaction id is in the format of 'shard.realm.num-ssssssssss.nnnnnnnnn'
+    // The transaction id is in the format of 'shard.realm.num-ssssssssss-nnnnnnnnn'
     // convert it in shard, realm, num and nanoseconds parameters
-    let txIdMatches = req.params.id.match(/(\d+)\.(\d+)\.(\d+)-(\d*)-(\d*)/);
+    let txIdMatches = req.params.id.match(/(\d+)\.(\d+)\.(\d+)-(\d{10})-(\d{9})/);
     if (txIdMatches === null || txIdMatches.length != 6) {
         logger.info(`getOneTransaction: Invalid transaction id ${req.params.id}`);
         res.status(404)

--- a/rest-api/utils.js
+++ b/rest-api/utils.js
@@ -21,6 +21,14 @@
 const math = require('mathjs');
 const config = require('./config.js');
 
+/**
+ * Check if the given number is numeric 
+ * @param {String} n Number to test
+ * @return {Boolean} true if n is numeric, false otherwise
+ */
+function isNumeric(n) {
+    return (!isNaN(parseFloat(n)) && isFinite(n));
+}
 
 /**
  * Split the account number into shard, realm and num fields. 
@@ -36,11 +44,15 @@ const parseEntityId = function (acc) {
 
     const aSplit = acc.split(".");
     if (aSplit.length == 3) {
-        ret.shard = aSplit[0];
-        ret.realm = aSplit[1];
-        ret.num = aSplit[2]
+        if (isNumeric(aSplit[0]) && isNumeric(aSplit[1]) && isNumeric(aSplit[2])) {
+            ret.shard = aSplit[0];
+            ret.realm = aSplit[1];
+            ret.num = aSplit[2]
+        }
     } else if (aSplit.length == 1) {
-        ret.num = acc;
+        if (isNumeric(acc)) {
+            ret.num = acc;
+        }
     }
     return (ret);
 }
@@ -121,6 +133,13 @@ const parseComparatorSymbol = function (fields, valArr, type = null, valueTransl
                         let ts = '' + seconds + (nanos + '000000000').substring(0, 9);
                         fquery += '(' + f + ' ' + opsMap[op] + ' ?) ';
                         vals.push(ts);
+                    } else if (type === 'balance') {
+                        if (isNumeric(val)) {
+                            fquery += '(' + f + ' ' + opsMap[op] + ' ?) ';
+                            vals.push(val);
+                        } else {
+                            fquery += '(1=1)';
+                        }
                     } else {
                         fquery += '(' + f + ' ' + opsMap[op] + ' ?) ';
                         vals.push(val);
@@ -278,8 +297,12 @@ const convertMySqlStyleQueryToPostgress = function (sqlQuery, sqlParams) {
  * @return {String} next Fully formed link to the next page
  */
 const getPaginationLink = function (req, isEnd, field, lastValue, order) {
-    const port = process.env.PORT;
-    const portquery = (Number(port) === 80) ? '' : (':' + port);
+    let urlPrefix;
+    if (process.env.PORT != undefined && process.env.INCLUDE_PATH_IN_NEXT_LINKS == 1) {
+        urlPrefix = req.protocol + '://' + req.hostname + ':' + process.env.PORT;
+    } else {
+        urlPrefix = '';
+    }
 
     var next = '';
 
@@ -322,7 +345,7 @@ const getPaginationLink = function (req, isEnd, field, lastValue, order) {
                 next += (next === '' ? '?' : '&') + q + '=' + v;
             }
         }
-        next = req.protocol + '://' + req.hostname + portquery + req.path + next;
+        next = urlPrefix + req.path + next;
     }
     return (next === '' ? null : next);
 }

